File: flink-utils/src/main/scala/io/github/ppdzm/utils/flink/scala/streaming/FlinkStreaming.scala
package io.github.ppdzm.utils.flink.scala.streaming

import io.github.ppdzm.utils.flink.scala.common.CheckpointConfiguration
import io.github.ppdzm.utils.universal.alert.Alerter
import io.github.ppdzm.utils.universal.base.Logging
import org.apache.flink.api.common.restartstrategy.RestartStrategies.RestartStrategyConfiguration
import org.apache.flink.streaming.api.functions.source.SourceFunction


/**
 * Created by Stuart Alex on 2021/4/13.
 */
trait FlinkStreaming[T] {
    protected lazy val logging = new Logging(getClass)
    /**
     * Checkpoint配置
     */
    protected lazy val checkpointConfiguration: CheckpointConfiguration = null
    /**
     * 应用名称
     */
    protected val applicationName: String
    /**
     * 告警器
     */
    protected val alerter: Alerter
    /**
     * 是否启用Checkpoint
     */
    protected val checkpointEnabled: Boolean
    /**
     * Flink应用重启配置
     */
    protected val restartStrategyConfiguration: RestartStrategyConfiguration
    /**
     * 数据源方法
     */
    protected val dataSource: SourceFunction[T]
    /**
     * 命令行参数
     */
    private var _args: Array[String] = _

    def main(args: Array[String]): Unit = {
        try {
            _args = args
            Logging.setLogging2Stdout(true)
            start()
            this.logging.logInfo(s"start execute application $applicationName")
        } catch {
            case e: Exception =>
                alerter.alert("", applicationName, e);
        }
    }

    protected def args: Array[String] = _args

    protected def start()
}

File: flink-utils/src/main/java/io/github/ppdzm/utils/flink/streaming/FlinkStreaming.java
package io.github.ppdzm.utils.flink.streaming;

import io.github.ppdzm.utils.flink.common.CheckpointConfiguration;
import io.github.ppdzm.utils.flink.streaming.config.FlinkStreamingConfig;
import io.github.ppdzm.utils.universal.alert.Alerter;
import io.github.ppdzm.utils.universal.base.Logging;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.source.SourceFunction;

import java.io.Serializable;

/**
 * @author Created by Stuart Alex on 2021/5/8.
 */
@Data
@EqualsAndHashCode()
@NoArgsConstructor
public abstract class FlinkStreaming<T> implements Serializable {
    private transient Logging logging = new Logging(getClass());
    /**
     * 应用名称
     */
    private String applicationName;
    /**
     * 告警器
     */
    private Alerter alerter;
    /**
     * Checkpoint配置
     */
    private CheckpointConfiguration checkpointConfiguration;
    private FlinkStreamingConfig flinkStreamingConfig;

    public FlinkStreaming(FlinkStreamingConfig flinkStreamingConfig, Alerter alerter) throws Exception {
        this();
        this.applicationName = flinkStreamingConfig.applicationName;
        this.alerter = alerter;
        this.checkpointConfiguration = flinkStreamingConfig.getCheckpointConfiguration();
        this.flinkStreamingConfig = flinkStreamingConfig;
    }

    public FlinkStreaming(String applicationName, Alerter alerter, CheckpointConfiguration checkpointConfiguration, FlinkStreamingConfig flinkStreamingConfig) {
        this();
        this.applicationName = applicationName;
        this.alerter = alerter;
        this.checkpointConfiguration = checkpointConfiguration;
        this.flinkStreamingConfig = flinkStreamingConfig;
    }

    public void start(SourceFunction<T> sourceFunction) throws Exception {
        try {
            StreamExecutionEnvironment streamExecutionEnvironment = flinkStreamingConfig.getStreamExecutionEnvironment();
            DataStreamSource<T> dataStreamSource = streamExecutionEnvironment.addSource(sourceFunction);
            execute(dataStreamSource);
            streamExecutionEnvironment.execute(applicationName);
            this.logging.logInfo("start execute application " + applicationName);
        } catch (Exception e) {
            alerter.alert("", applicationName, e);
            throw e;
        }
    }

    /**
     * 执行对dataStreamSource进行的操作
     *
     * @param dataStreamSource 数据源
     */
    public abstract void execute(DataStreamSource<T> dataStreamSource);

}
